sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp

```diff
@@ -119,15 +120,37 @@ static sycl::detail::ESIMDEmuPluginOpaqueData *PiESimdDeviceAccess;
 // Single-entry cache for piPlatformsGet call.
 static pi_platform PiPlatformCache;
 // TODO/FIXME : Memory leak. Handle with 'piTearDown'.
-static sycl::detail::SpinLock *PiPlatformCacheMutex =
-    new sycl::detail::SpinLock;
+static std::mutex *PiPlatformCacheLock = new std::mutex;
 
-// Mapping between surface index and CM-managed surface
+// Mapping between surface index and surfaces from both Host and
+// CM. No need to serialize accesses to this map as the SurfaceIndex
+// generation for this map always generates unique key atomically and
+// generated keys are never re-used for another instance of _pi_mem object
+// REF : _pi_context::generateSurfaceIndex()
 static std::unordered_map<unsigned int, _pi_mem *> *PiESimdSurfaceMap =
     new std::unordered_map<unsigned int, _pi_mem *>;
-// TODO/FIXME : Memory leak. Handle with 'piTearDown'.
-static sycl::detail::SpinLock *PiESimdSurfaceMapLock =
-    new sycl::detail::SpinLock;
+
+pi_result _pi_mem::addMapping(void *MappedTo) {
```

Comment:
@romanovvlad , @smaslov-intel , I imported `addMapping/removeMapping` from level_zero PI implementation. Although those functions are removed recently in level_zero, sanity checks still remain there. I have following questions

- Are these sanity checks (duplication check for adding, invalid unmapping check for removing) are necessary? If so why?
- I replaced `unordered_map` with `stack` per @romanovvlad 's suggestion while level_zero implementation still uses `unordered_map`. Which one is correct? In case of `stack`-implementation, is it assumed that Map/Unmap calls are thread-safe?

Reply:
>> Are these sanity checks (duplication check for adding, invalid unmapping check for removing) are necessary? If so why?

My understanding is that two calls with the same `MappedTo` are legal. In the `removeMapping` I would suggest having an assert checking that `MappedTo == peak`, though runtime check(as it is right now) is also OK.

>> I replaced unordered_map with stack per @romanovvlad 's suggestion while level_zero implementation still uses unordered_map. Which one is correct? In case of stack-implementation, is it assumed that Map/Unmap calls are thread-safe?

PI Map/Unmap might be called by SYCL RT from several threads at the same time without any synchronization, so the operations with `Mappings` should be guarded by a mutex. I suggested `std::stack` as a more simple and cheap structure for map/unmap operations because, IIUC, unmap operation should always "unmap" the latest map operation.

Reply:
`unordered_map` is applied with TODO comments for changing to `stack` type.